drivers/sensor/ds18b20/ds18b20.c

```diff
@@ -0,0 +1,254 @@
+{
+	int ret;
+	struct ds18b20_data *data = dev->data;
+	const struct device *bus = ds18b20_bus(dev);
+
+	ret = w1_reset_select(bus, &data->config);
+	if (ret != 0) {
+		return ret;
+	}
+
+	return w1_write_byte(bus, DS18B20_CMD_CONVERT_T);
+}
+
+/* Read power supply status from the sensor. */
+static int ds18b20_read_power_supply(const struct device *dev)
+{
+	struct ds18b20_data *data = dev->data;
+	const struct device *bus = ds18b20_bus(dev);
+	int ret;
+	uint8_t cmd = DS18B20_CMD_READ_POWER_SUPPLY;
+
+	ret = w1_write_read(bus, &data->config, &cmd, 1, NULL, 0);
+	if (ret != 0) {
+		return ret;
+	}
+
+	ret = w1_read_bit(bus);
+	if (ret < 0) {
+		return ret;
+	}
+	data->power_supply = (bool)ret;
+	return 0;
+}
+
+/*
+ * Write resolution into configuration struct,
+ * but don't write it to the sensor yet.
+ */
+static void ds18b20_set_resolution(const struct device *dev, uint8_t resolution)
+{
+	struct ds18b20_data *data = dev->data;
+
+	data->scratchpad.config &= ~DS18B20_RESOLUTION_MASK;
+	data->scratchpad.config |= DS18B20_RESOLUTION(resolution);
+}
+
+static int ds18b20_sample_fetch(const struct device *dev,
+				enum sensor_channel chan)
+{
+	const struct ds18b20_config *cfg = dev->config;
+	struct ds18b20_data *data = dev->data;
+	int status;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL ||
+			chan == SENSOR_CHAN_AMBIENT_TEMP);
+
+	if (!data->lazy_loaded) {
+		status = ds18b20_configure(dev);
+		if (status < 0) {
+			return status;
+		}
+		data->lazy_loaded = true;
+	}
+
+	status = ds18b20_temperature_convert(dev);
+	if (status < 0) {
+		LOG_DBG("W1 fetch error");
+		return status;
+	}
+	k_msleep(measure_wait_ms[DS18B20_RESOLUTION_INDEX(cfg->resolution)]);
+	return ds18b20_read_scratchpad(dev, &data->scratchpad);
+}
+
+static int ds18b20_channel_get(const struct device *dev,
+			       enum sensor_channel chan,
+			       struct sensor_value *val)
+{
+	struct ds18b20_data *data = dev->data;
+
+	if (chan != SENSOR_CHAN_AMBIENT_TEMP) {
+		return -ENOTSUP;
+	}
+
+	ds18b20_temperature_from_raw((uint8_t *)&data->scratchpad.temp, val);
+	return 0;
+}
+
+static int ds18b20_configure(const struct device *dev)
+{
+	const struct ds18b20_config *cfg = dev->config;
+	struct ds18b20_data *data = dev->data;
+
+	/* In single drop configurations the rom can be read from device */
+	if (w1_get_peripheral_count(cfg->bus) == 1) {
+		if (w1_rom_to_uint64(&data->config.rom_id) == 0ULL) {
+			(void)w1_read_rom(cfg->bus, &data->config.rom_id);
+		}
+	} else if (w1_rom_to_uint64(&data->config.rom_id) == 0ULL) {
+		LOG_DBG("nr: %d", w1_get_peripheral_count(cfg->bus));
+		LOG_ERR("ROM required, because multiple devices are on the bus");
+		return -EINVAL;
+	}
+
+	if ((cfg->family != 0) && (cfg->family != data->config.rom_id.family)) {
+		LOG_ERR("Found 1-Wire device is not a DS18B20");
+		return -EINVAL;
+	}
+
+	if (w1_reset_bus(cfg->bus) <= 0) {
+		LOG_ERR("No 1-Wire devices connected");
+		return -EINVAL;
+	}
```

Comment:
Shouldn't here return value be `-ENODEV`, that would be in line with return value from `w1_read_rom` that returns `-ENODEV` in exactly the same, bus reset, case.

Reply:
I do agree that this should be in line with the others and have applied the suggestion.
Furthermore, I have moved this check to the beginning of ds18b20_configure.
Additionally, I have changed the locking in temperature_convert() and read_power_supply().